presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeIdenticalWindows.java

```diff
@@ -0,0 +1,148 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.optimizations;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.metadata.Signature;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.SimplePlanRewriter;
+import com.facebook.presto.sql.planner.plan.WindowNode;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkState;
+
+/**
+ * Merge together the functions in WindowNodes that have identical WindowNode.Specifications.
+ * For example:
+ *
+ * OutputNode
+ * `--...
+ *    `--WindowNode(Specification: A, Functions: [sum(something)])
+ *       `--WindowNode(Specification: B, Functions: [sum(something)])
+ *          `--WindowNode(Specification: A, Functions: [avg(something)])
+ *             `--...
+ *
+ * Will be transformed into
+ *
+ * OutputNode
+ * `--...
+ *    `--WindowNode(Specification: A, Functions: [sum(something), avg(something)])
+ *       `--WindowNode(Specification: B, Functions: [sum(something)])
+ *          `--...
+ *
+ * This will NOT merge the functions in WindowNodes that have identical WindowNode.Specifications,
+ * but have a node between them that is not a WindowNode.
+ * In the following example, the functions in the WindowNodes with specification `A' will not be
+ * merged into a single WindowNode.
+ *
+ * OutputNode
+ * `--...
+ *    `--WindowNode(Specification: A, Functions: [sum(something)])
+ *       `--WindowNode(Specification: B, Functions: [sum(something)])
+ *          `-- ProjectNode(...)
+ *             `--WindowNode(Specification: A, Functions: [avg(something)])
+ *                `--...
+ */
+public class MergeIdenticalWindows
+        implements PlanOptimizer
+{
+    public PlanNode optimize(PlanNode plan,
+            Session session,
+            Map<Symbol, Type> types,
+            SymbolAllocator symbolAllocator,
+            PlanNodeIdAllocator idAllocator)
+    {
+        return SimplePlanRewriter.rewriteWith(new Rewriter(), plan, new Context());
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<MergeIdenticalWindows.Context>
+    {
+        @Override
+        public PlanNode visitWindow(
+                WindowNode node,
+                RewriteContext<MergeIdenticalWindows.Context> context)
+        {
+            Context mergeContext = context.get();
```

Comment:
Since this optimizer is potentially changing the order of the window functions potentially, let's make sure that things that may be order dependent are not yet in place. We should add the following checks:

```java
checkState(!node.getHashSymbol().isPresent(), "MergeIdenticalWindows should be run before HashGenerationOptimizer");
checkState(node.getPrePartitionedInputs().isEmpty() && node.getPreSortedOrderPrefix() == 0, "MergeIdenticalWindows should be run before AddExchanges");
```